### ISSUE TITLE
Safe to use setGlobal anywhere?

### ISSUE THREAD

**Issue by andy9775** [CharlesStover/reactn#27]
Hey, really cool library.

I have a quick question though. Is using `setGlobal` inside components safe? Specifically functional components. I'm using the following pattern for instance:
```typescript
import React from 'react';
import { setGlobal, useGlobal } from 'reactn';

interface InputProps {
  stateKey: string; // state identifier
  label: string;
}

export default (props: InputProps) => {
  const [value, setValue] = useGlobal(props.stateKey);
  if (value === undefined) {
    setGlobal({ [props.stateKey]: '' });
  }

  return (
    <div>
      <div>{props.label}</div>
      <input
        value={value}
        onChange={e => {
          e.preventDefault();
          setValue(e.target.value);
        }}
      />
    </div>
  );
};


```
based off of the example code that uses classes and calls `setGlobal` in `componentDidMount`. 

Finally, I tried calling `setGlobal` inside of `useEffect` to no effect hence the need for wrapping `setGlobal` in an if statement. Is this a bug or intentional?

**Comment by CharlesStover** [comment]
The `setGlobal` helper function is _safe_ to use, but I'm not sure I'd recommend it. I can't think of a reason it would ever cause an error to use it like this, but it isn't the design pattern; it just coincidentally works fine. I can't guarantee it will always work equivalently in future versions.

This may be closer to what you are looking for:

```JavaScript
const [ value, setValue ] = useGlobal(props.stateKey);
if (value === undefined) {
  setValue('');
  return null; // Do not render until a value exists.
}
```

**Comment by andy9775** [comment]
Hey @CharlesStover, thanks for taking the time to dive so deep into - truly appreciate it.

I haven't had time to explore either option yet, but looking at your proposals it looks like I can make them work - option 2 looks like it will work without too many changes.

<hr/>

The goal I'm trying to achieve is to be able to do something like this for index pages (lists):
```jsx
<ListWrapper>
    <FilterGroup><SomeFilter/> ... </FilterGroup>
    <List data={...} />
</ListWrapper>
```
Then internally the list wrapper contains all the logic to wire up the filter group filters and list pagination logic into global state. Hence, once a user clicks on a list item and navigates away from the index page we can keep the pagination and filters after the index page unmounts. Then when going back we've persisted the data so the user goes back to where they left off. My current approach is to store this on redux and have a reducer to handle everything. I'd like to get rid of the boilerplate of having to wire everything up (redux, initial state, pagination logic, etc.) per list.

**Comment by CharlesStover** [comment]
ReactN can definitely help you decrease your boilerplate. I have tested the code I provided for the second example to work, if you want to use it as a basis for your logic.

In response to my edit, I don't think I will implement force updating for non-existing keys. A component setting a non-existent key does not _necessarily_ mean it should re-render. It does in your use case. I don't think it does in _all_ use cases, so I don't want that functionality impacting other scenarios.

```JavaScript
const X = () => {
  const [ yRenders ] = useGlobal('yRenders');
  return 'Y has rendered ' + yRenders + ' times!';
};
const Y = () => {
  const incrementYRenders = useGlobal(global => ({
    yRenders: (global.yRenders || 0) + 1
  }));
  incrementYRenders();
  return 'Y';
};
```
In the above example, while the Y component _sets_ `yRenders`, it should not re-render in response to the change.
